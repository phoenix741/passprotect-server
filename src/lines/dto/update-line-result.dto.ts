import { createUnionType } from 'type-graphql';
import { Errors } from '../../shared/dto/errors.dto';
import { WalletLine } from './wallet-line.dto';

export const UpdateLineResultUnion = createUnionType({
  name: 'UpdateLineResult',
  description: 'Result of creation or update of the line',
  types: [WalletLine, Errors],
  resolveType: (obj: WalletLine | Errors): string => {
    const hasError = (tbd: WalletLine | Errors): tbd is Errors => !!(tbd as Errors).errors;
    if (hasError(obj)) {
      return 'Errors';
    }
    return 'WalletLine';
  },
});

export type UpdateLineResult = typeof UpdateLineResultUnion;
