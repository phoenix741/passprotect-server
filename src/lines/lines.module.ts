import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { LineSchema } from './schemas/line.schema';
import { LinesService } from './lines.service';
import { SharedModule } from '../shared/shared.module';
import { TransactionsModule } from '../transactions/transactions.module';

@Module({
  imports: [SharedModule, MongooseModule.forFeature([{ name: 'Line', schema: LineSchema }]), TransactionsModule],
  providers: [LinesService],
  exports: [LinesService],
})
export class LinesModule {}
