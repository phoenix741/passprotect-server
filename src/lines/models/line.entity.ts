import { Document } from 'mongoose';
import { EncryptedContent } from '../../shared/interfaces/encrypted-content.interface';
import { LineTypeEnum } from '../../shared/interfaces/line-type-enum.interface';
import { ObjectID } from 'bson';

export interface LineEntity extends Document {
  _id: ObjectID;
  type: LineTypeEnum;
  label: string;
  group: string;
  logo?: Buffer;
  encryption: EncryptedContent;
  user: string;
  createdAt: Date;
  updatedAt: Date;
  _rev: number;
}
