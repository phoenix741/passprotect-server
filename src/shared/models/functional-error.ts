import { HttpException } from '@nestjs/common';

export class FunctionalError extends Error {
  field?: string;

  constructor(field?: string, message?: string) {
    super(message);

    this.field = field;
  }

  toGraphQL(defaultLabel = 'label') {
    return {
      errors: [{
        fieldName: this.field || defaultLabel,
        message: this.message,
      }],
    };
  }
}

export function toFunctionalError(error: Error): FunctionalError {
  if (error instanceof FunctionalError) {
    return error;
  }
  if (error instanceof HttpException) {
    return new FunctionalError(undefined, error.message.message);
  }
  return new FunctionalError(undefined, error.message);
}
