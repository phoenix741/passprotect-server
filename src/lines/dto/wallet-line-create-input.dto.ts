import { InputType, Field, ID } from 'type-graphql';
import { EncryptedBlockInput } from '../../shared/dto/encrypted-block-input.dto';
import { LineTypeEnum } from '../../shared/interfaces/line-type-enum.interface';

@InputType({ description: 'Wallet line used for creation' })
export class WalletLineCreateInput {
  @Field(type => LineTypeEnum, {
    description: 'Type of the line (text, password, card)',
  })
  type: LineTypeEnum;

  @Field({ description: 'Label of the wallet line' })
  label: string;

  @Field({ description: 'Group of the wallet line' })
  group: string;

  @Field({ description: 'Logo of the line', nullable: true })
  logo?: string;

  @Field(type => EncryptedBlockInput, {
    description: 'Encrypted content of the line (encrypted client side)',
  })
  encryption: EncryptedBlockInput;
}
