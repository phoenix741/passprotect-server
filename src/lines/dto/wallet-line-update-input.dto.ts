import { InputType, Field } from 'type-graphql';
import { ObjectID } from 'bson';
import { LineTypeEnum } from '../../shared/interfaces/line-type-enum.interface';
import { EncryptedBlockInput } from '../../shared/dto/encrypted-block-input.dto';

@InputType({ description: 'Wallet line used for update' })
export class WalletLineUpdateInput {
  @Field(type => ObjectID, {
    description: 'Id of the line, if not filled application create a new line',
    nullable: true,
  })
  _id: ObjectID;

  @Field(type => LineTypeEnum, {
    description: 'Type of the line (text, password, card)',
    nullable: true,
  })
  type?: LineTypeEnum;

  @Field({ description: 'Label of the wallet line', nullable: true })
  label?: string;

  @Field({ description: 'Group of the wallet line', nullable: true })
  group?: string;

  @Field({ description: 'Logo of the line', nullable: true })
  logo?: string;

  @Field(type => EncryptedBlockInput, {
    description: 'Encrypted content of the line (encrypted client side)',
    nullable: true,
  })
  encryption?: EncryptedBlockInput;
}
