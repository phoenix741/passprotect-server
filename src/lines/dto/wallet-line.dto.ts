import { Field, ObjectType, ID, Int } from 'type-graphql';
import { EncryptedBlock } from '../../shared/dto/encrypted-block.dto';
import { LineTypeEnum } from '../../shared/interfaces/line-type-enum.interface';
import { ObjectID } from 'bson';

@ObjectType({ description: 'Wallet' })
export class WalletLine {
  @Field(type => ObjectID, { description: 'Id of the line' })
  _id: ObjectID;

  @Field(type => LineTypeEnum, { description: 'Type of the line (text, password, card)' })
  type: LineTypeEnum;

  @Field({ description: 'Label of the wallet line' })
  label: string;

  @Field({ description: 'Group of the wallet line', nullable: true })
  group?: string;

  @Field(type => Buffer, { description: 'Logo of the line', nullable: true })
  logo?: Buffer;

  @Field(type => EncryptedBlock, { description: 'Encrypted content of the line (encrypted client side)' })
  encryption: EncryptedBlock;

  @Field(type => Date, { description: 'Created date of the line', nullable: true })
  createdAt?: Date;

  @Field(type => Date, { description: 'Last updated date of the line', nullable: true })
  updatedAt?: Date;

  @Field(type => Int, { description: 'Revision (modification count) of the line' })
  _rev?: number;
}
