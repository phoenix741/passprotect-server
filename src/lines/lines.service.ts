import { Model } from 'mongoose';
import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { LineToCreate, LineToUpdate } from './models/line.models';
import { LineEntity } from './models/line.entity';
import { ObjectID } from 'bson';
import { TransactionsService } from '../transactions/transactions.service';
import { TransactionEntity } from '../transactions/models/transaction.entity';
import { TransactionTypeEnum } from '../shared/interfaces/transaction-type-enum.interface';

interface IFindAllFilter {
  user?: string;
}

@Injectable()
export class LinesService {
  constructor(
    @InjectModel('Line') private readonly lineModel: Model<LineEntity>,
    private readonly transactionService: TransactionsService,
  ) {}

  async findAll(userId: string): Promise<LineEntity[]> {
    if (userId) {
      const filter: IFindAllFilter = { user: userId };
      const sort = { group: 1, label: 1 };
      return await this.lineModel
        .find(filter)
        .sort(sort)
        .exec();
    }
    return [];
  }

  async findById(id: ObjectID) {
    return await this.lineModel.findById(id);
  }

  async createLine(line: LineToCreate): Promise<LineEntity> {
    const lineToCreate = new this.lineModel(line);
    const newLine = await lineToCreate.save();
    await this.transactionService.createTransaction(
      TransactionTypeEnum.line,
      null,
      newLine,
    );
    return newLine;
  }

  async updateLine(
    originalLine: LineEntity,
    line: LineToUpdate,
  ): Promise<LineEntity> {
    const { _id, ...cleanLine } = line;

    const oldLine = new this.lineModel(originalLine.toObject());
    originalLine.set(cleanLine);
    const newLine = await originalLine.save();

    await this.transactionService.createTransaction(
      TransactionTypeEnum.line,
      oldLine,
      newLine,
    );
    return newLine;
  }

  async removeLine(id: ObjectID): Promise<TransactionEntity> {
    const oldLine = await this.lineModel.findById(id);
    await this.lineModel.deleteOne({ _id: id });
    return this.transactionService.createTransaction(
      TransactionTypeEnum.line,
      oldLine,
      null,
    );
  }

  async getGroups(userId: string): Promise<string[]> {
    const filter = { user: userId };
    return this.lineModel.distinct('group', filter);
  }
}
