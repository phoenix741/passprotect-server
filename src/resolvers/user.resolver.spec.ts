import { Test } from '@nestjs/testing';
import { UserResolver } from './user.resolver';
import { UsersService } from 'src/users/users.service';
import { LinesService } from 'src/lines/lines.service';
import { AuthorizationService } from 'src/shared/services/authorization.service';
import { User } from 'src/users/dto/user.dto';
import { LineEntity } from 'src/lines/models/line.entity';
import { RegistrationUserInput } from 'src/users/dto/registration-user.dto';
import { FunctionalError } from 'src/shared/models/functional-error';

describe('UserResolver', () => {
  let userResolver: UserResolver;
  let userService;
  let linesService;
  let authorizationService;

  const user: User = new User();
  const lines: LineEntity[] = [
    { _id: 'lineId' } as any as LineEntity,
  ];

  beforeEach(async () => {
    userService = {
      findById: jest.fn(),
      registerUser: jest.fn(),
    };
    linesService = {
      findAll: jest.fn(),
    };
    authorizationService = {
      checkPermission: jest.fn(),
    };
    user._id = 'username';

    const module = await Test.createTestingModule({
      providers: [
        UserResolver,
        {
          provide: UsersService,
          useValue: userService,
        },
        {
          provide: LinesService,
          useValue: linesService,
        },
        {
          provide: AuthorizationService,
          useValue: authorizationService,
        },
      ],
    })
    .compile();

    userResolver = module.get<UserResolver>(UserResolver);
  });

  describe('user', () => {
    it('success', async () => {
      userService.findById.mockImplementation(() => user);
      expect(await userResolver.user('username', user)).toMatchSnapshot();
      expect(authorizationService.checkPermission).toHaveBeenCalledWith(user, 'username');
      expect(userService.findById).toHaveBeenCalledWith('username');
    });
  });

  describe('lines', () => {
    it('success', async () => {
      linesService.findAll.mockImplementation(() => lines);
      expect(await userResolver.lines(user, user)).toMatchSnapshot();
      expect(authorizationService.checkPermission).toHaveBeenCalledWith(user, 'username');
      expect(linesService.findAll).toHaveBeenCalledWith('username');
    });
  });

  describe('registerUser', () => {
    it('success', async () => {
      const input = new RegistrationUserInput();
      input._id = 'registeredUsername';
      input.password = 'password';
      userService.registerUser.mockImplementation(() => user);
      expect(await userResolver.registerUser(input)).toMatchSnapshot();
      expect(userService.registerUser).toHaveBeenCalledWith(input);
    });

    it('error', async () => {
      const input = new RegistrationUserInput();
      input._id = 'registeredUsername';
      input.password = 'password';
      userService.registerUser.mockImplementation(() => {
        throw new FunctionalError('myfield', 'mymessage');
      });
      expect(await userResolver.registerUser(input)).toMatchSnapshot();
      expect(userService.registerUser).toHaveBeenCalledWith(input);
    });
  });
});
