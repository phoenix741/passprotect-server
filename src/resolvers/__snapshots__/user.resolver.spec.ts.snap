// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserResolver lines success 1`] = `
Array [
  Object {
    "_id": "lineId",
  },
]
`;

exports[`UserResolver registerUser error 1`] = `
Object {
  "errors": Array [
    Object {
      "fieldName": "myfield",
      "message": "mymessage",
    },
  ],
}
`;

exports[`UserResolver registerUser success 1`] = `
User {
  "_id": "username",
}
`;

exports[`UserResolver user success 1`] = `
User {
  "_id": "username",
}
`;
